aiothrift/pool.py

```diff
@@ -182,84 +177,27 @@ def fill_free(self, *, override_min):
+                    conn = await self._create_new_connection()
                     self._pool.append(conn)
                 finally:
                     self._acquiring -= 1
 
     def _create_new_connection(self):
         return create_connection(self._service, self._address,
-                                 loop=self._loop, timeout=self._timeout)
+                                 timeout=self._timeout)
 
-    @asyncio.coroutine
-    def _notify_conn_returned(self):
-        with (yield from self._cond):
+    async def _notify_conn_returned(self):
+        async with self._cond:
             self._cond.notify()
 
-    def __enter__(self):
-        raise RuntimeError(
-            "'yield from' should be used as a context manager expression")
-
-    def __exit__(self, *args):
-        pass  # pragma: nocover
-
-    def __iter__(self):
-        # this method is needed to allow `yield`ing from pool
-        conn = yield from self.acquire()
-        return _ConnectionContextManager(self, conn)
-
-    if PY_35:
-        def __await__(self):
-            # To make `with await pool` work
-            conn = yield from self.acquire()
-            return _ConnectionContextManager(self, conn)
-
-        def get(self):
-            """
-            Return async context manager for working with connection::
-
-                async with pool.get() as conn:
-                    await conn.get(key)
-            """
-            return _AsyncConnectionContextManager(self)
-
-
-class _ConnectionContextManager:
-    __slots__ = ('_pool', '_conn')
-
-    def __init__(self, pool, conn):
-        self._pool = pool
-        self._conn = conn
-
-    def __enter__(self):
-        return self._conn
-
-    def __exit__(self, exc_type, exc_value, tb):
-        try:
-            self._pool.release(self._conn)
-        finally:
-            self._pool = None
-            self._conn = None
-
-
-if PY_35:
-    class _AsyncConnectionContextManager:
-
-        __slots__ = ('_pool', '_conn')
-
-        def __init__(self, pool):
-            self._pool = pool
-            self._conn = None
-
-        @asyncio.coroutine
-        def __aenter__(self):
-            self._conn = yield from self._pool.acquire()
-            return self._conn
-
-        @asyncio.coroutine
-        def __aexit__(self, exc_type, exc_value, tb):
-            try:
-                self._pool.release(self._conn)
-            finally:
-                self._pool = None
-                self._conn = None
+    async def __aenter__(self):
+        if self.closed:
+            raise PoolClosedError('cannot acquire a connection from a closed pool')
+        if acquired_connection.get(None) is not None:
+            raise RuntimeError('cannot acquire a connection if you already have one inside the same task')
+        _conn = await self.acquire()
+        acquired_connection.set(_conn)
+        return _conn
+
+    async def __aexit__(self, *exc_info):
+        self.release(acquired_connection.get())
```

Comment:
I thinks it's better to put all logic to `acquire` and `release`  and just make `__aenter__` and `__aexit__` a context wrapper above.